projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php

```diff
@@ -0,0 +1,152 @@
+<?php
+/**
+ * Module: Jetpack Google Fonts
+ *
+ * @package automattic/jetpack
+ */
+
+use Automattic\Jetpack\Fonts\Global_Styles_Webfonts_Introspector;
+
+/**
+ * Manages Google fonts registration, introspection and enqueueing.
+ */
+class Jetpack_Google_Fonts {
+	/**
+	 * The Google fonts that we're registered after
+	 * running the `jetpack_google_fonts_list`.
+	 *
+	 * @var array
+	 */
+	private $registered_google_fonts = array();
+
+	/**
+	 * Hook into WordPress to register
+	 * and enqueue webfonts.
+	 */
+	public function __construct() {
+		$this->register_google_fonts_provider();
+
+		add_action( 'after_setup_theme', array( $this, 'register_google_fonts' ) );
+
+		/**
+		 * We are already enqueueing all registered fonts by default when loading the block editor,
+		 * so we only need to scan for webfonts when browsing as a guest.
+		 */
+		if ( ! is_admin() ) {
+			add_action( 'wp_loaded', array( $this, 'enqueue_global_styles_google_fonts' ) );
+			add_filter( 'pre_render_block', array( $this, 'scan_block_for_google_fonts' ), 10, 2 );
+		}
+	}
+
+	/**
+	 * Register the provider for Google fonts registration.
+	 */
+	private function register_google_fonts_provider() {
+		if ( ! function_exists( 'wp_register_webfont_provider' ) ) {
+			return;
+		}
+
+		wp_register_webfont_provider( 'jetpack-google-fonts', '\Automattic\Jetpack\Fonts\Google_Fonts_Provider' );
+	}
+
+	/**
+	 * Register Google fonts.
+	 */
+	public function register_google_fonts() {
+		if ( ! function_exists( 'wp_register_webfonts' ) ) {
+			return;
+		}
+
+		/**
+		 * Curated list of Google Fonts.
+		 *
+		 * @module google-fonts
+		 *
+		 * @since 10.8
+		 *
+		 * @param array $fonts_to_register Array of Google Font names to register.
+		 */
+		$fonts_to_register = apply_filters( 'jetpack_google_fonts_list', JETPACK_GOOGLE_FONTS_LIST );
+
+		foreach ( $fonts_to_register as $font_family ) {
+			$font_family_slugs = wp_register_webfonts(
+				array(
+					array(
+						'font-family'  => $font_family,
+						'font-weight'  => '100 900',
+						'font-style'   => 'normal',
+						'font-display' => 'fallback',
+						'provider'     => 'jetpack-google-fonts',
+					),
+					array(
+						'font-family'  => $font_family,
+						'font-weight'  => '100 900',
+						'font-style'   => 'italic',
+						'font-display' => 'fallback',
+						'provider'     => 'jetpack-google-fonts',
+					),
+				)
+			);
+
+			if ( empty( $font_family_slugs ) ) {
+				// Fonts were not registered.
+				continue;
```

Comment:
Could this lead to errors being silently hidden?

Reply:
Not really -- we're always emitting errors in the front-end so they'll be noticed.

Reply:
Hmmm...I might be missing something but here's the flow of logic I have in my head ( I might be mistaken about this! )

- We'd attempt to register webfonts
- `register_webfont` method would return `false`
- `wp_register_webfonts` would fail and return an empty array of `font_family_slugs`
- We wouldn't enqueue any fonts because `registered_google_fonts` would be empty
- The frontend would silently fallback to some system or default font

When would errors be emitted on the frontend?

Reply:
The `wp_register_webfonts` method runs for every font family declared in the list, so there'd be multiple runs and not a single one.

And yes, that's mostly right. One thing, though, is that the user/developer would be notified after `register_webfont method would return false` on the front-end, so they'd have a clue on how to fix what went wrong.

Reply:
> register_webfont method would return false on the front-end, so they'd have a clue on how to fix what went wrong

Ah I see -- that's what I was missing. When you have time, would you mind explaining how returning false would notify the user / dev in the frontend? The only way I can see this working is with error logging.

Reply:
It's not for the front-end. Return `false` is more about conditional action upon registration failure for example. I do agree, though, that we're lacking descriptive return values on failure when dealing with the back-end (i.e. `false` can mean multiple things). But that's not something to address in this PR.

Reply:
> But that's not something to address in this PR.

Agreed

Reply:
Oh, my bad.

What I tried to say was that a notice is displayed on the front end, so the user can take action.